### DISCUSSION TITLE
User emails are case sensative?

### DISCUSSION THREAD

**Discussion by calumk** [pocketbase/pocketbase#756] [Q&A]
The User (and maybe admin?) Emails seem to be case sensative auth?

Passwords should be case sensative, but should emails?

### Steps to produce: 
1) Create user : `USER@email.com | p4ssword `

2) Try to auth : `user@email.com | p4ssword` = Fail 400

2) Try to auth : `USER@email.com | p4ssword` = Succsess

**Comment by ganigeorgiev** [comment]
I'm not sure. Technically, its up to the host how they will treat the local part (aka. before the `@domain`). Gmail for example allows appending special symbols to the same email address and still treating is as a single mail account.

From this RFC - https://www.rfc-editor.org/rfc/rfc5321#section-2.4:

> SMTP implementations MUST take care to preserve the case of mailbox local-parts. In particular, for some hosts, the user "smith" is different from the user "Smith".

Whether the same should be applied when checking for uniqueness in the database - I'm not really sure. If we want to be compliant with the above, then the current case sensitive address behavior is correct.

_I'll move the issue to "Q&A" discussion._

**Reply by calumk** [reply]
This is interesting!
Ok, so the RFC shows clearly that it should be treated as case-sensative.
 - But can we find a single example of a system that implements in this way?

Would anyone really want 
`user@email.com `and `USER@email.com `as two seperate pocketbase users?

**Reply by ganigeorgiev** [reply]
> Would anyone really want user@email.com and USER@email.com as two seperate pocketbase users?

The same could be asked for `test+1@gmail.com` and `test+2@gmail.com` or `a.b@gmail.com` and `ab@gmail.com`, etc.
It is hard to tell really because it depends on the mail host. At the end, making it case insensitive will not stop your users to setup their own mail server and have all `*@example.com` addresses to point to the same mail inbox. I guess most mail providers allow those special characters and treat the local part as case insensitive to improve the deliverability in case of a typo or something similar.

**Reply by ganigeorgiev** [reply]
You can enforce only lowercase emails either server-side with a before save hook or simply only with js on the client side (at the end the email case insensitivity is more of "convenience" feature for your users to prevent confusion and minimize the failed login attempts).

I'm not sure if it will be very useful to change the current behavior. Adding a configuration option only for this doesn't feel right to me either since it may clutter the UI unnecessary.

**Reply by bouchtaoui-dev** [reply]
I have this issue too and I agree with @calumk, but I do understand your decision too.
I guess making email address lowercase on client side for registration as well as authentication would suffice for me.
I wonder if this is the same on firebase for example?
Anyway, I'm still very happy for this fantastic backend :D

**Comment by MrSharpp** [comment]
i have used abc@gmail.com & abc@Gmail.com
and now i have 2 accounts :-)

**Reply by ganigeorgiev** [reply]
Please read https://github.com/pocketbase/pocketbase/discussions/756#discussioncomment-3849199.

I may consider in the future making an exception for the domain part, but again you shouldn't treat this as some kindof protection since the local part could be anything (for example, `a.bc@gmail.com` would also point to the same inbox).

**Reply by MrSharpp** [reply]
so I need to make a wrapper for the pocket base API for the email part if I am using a pocket base in my app auth?

**Reply by ganigeorgiev** [reply]
Please read the comments in the above discussion.

If you want to enforce case sensitivity you can do this either client-side or server-side with a [before create/update hook](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks), but once again, don't treat this as some kind of protection. The local part can be anything and it is up to the mail provider whether it will represent a different user or not.

**Comment by benallfree** [comment]
I’ve never encountered this until now. As I see the options:

1. Do nothing. (Users get very confused why they can’t log in or even perform pw recovery)
2. Force to lowercase. (May affect deliverability and account verification)
3. Consider emails that vary only by case as duplicate and disallow account creation. Modify login and pw recovery to query as case insensitive. (May prevent some legitimate users from signing up using legitimate distinct email addresses)

I think option 3 is probably safest for us. It may in very rare cases mean that user Foo and user foo cannot both sign up, but that seems almost impossibly rare compared to the confusion created by case sensitivity.

**Comment by ganigeorgiev** [comment]
_Just for info in case someone stumble here, with PocketBase v0.25+ if you want to enforce case-insentitive email validations and lookups another option could be to can change the `email` field UNIQUE index from the collection options and specify the `COLLATE NOCASE` collation:_

```sql
-- by default NOCASE works only with ascii characters, for unicode you'll have to load the SQLite ICU extension
-- (ex. https://github.com/pocketbase/pocketbase/issues/6604#issuecomment-2727445264)
CREATE UNIQUE INDEX `__pb_users_auth__email_idx` ON `users` (`email` COLLATE NOCASE) WHERE `email` != ''
```
_(`FindAuthRecordByEmail` and other similar auth query methods will try to inspect the collection indexes and they will apply the same collation during search)_
